pkg/provision/gce.go

```diff
@@ -0,0 +1,192 @@
+package provision
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"google.golang.org/api/compute/v1"
+	"google.golang.org/api/option"
+)
+
+// GCEProvisioner holds a GCE compute service for compute related operations
+type GCEProvisioner struct {
+	computeService *compute.Service
+}
+
+// NewGCEProvisioner with an accessKeyFile
+func NewGCEProvisioner(ctx context.Context, accessKeyFileByte string) (*GCEProvisioner, error) {
+	computeSVC, err := compute.NewService(ctx, option.WithCredentialsJSON([]byte(accessKeyFileByte)))
+	return &GCEProvisioner{
+		computeService: computeSVC,
+	}, err
+}
+
+// Provision provisions a GCE instance and Firewall rules to allow the control-port
+func (p *GCEProvisioner) Provision(host BasicHost) (*ProvisionedHost, error) {
+	if host.Additional["zone"] == "" {
+		host.Additional["zone"] = "us-central1-a"
+	}
+
+	// auto-restart instance on failure
+	autoRestart := true
+
+	network := host.Additional["network"]
+
+	instance := &compute.Instance{
+		Name:        host.Name,
+		Description: "inlets exit node",
+		MachineType: fmt.Sprintf("zones/%s/machineTypes/%s", host.Additional["zone"], host.Plan),
+		Tags: &compute.Tags{
+			Items: []string{
+				"http-server",
+				"https-server",
+				"inlets",
+			},
+		},
+		Metadata: &compute.Metadata{
+			Items: []*compute.MetadataItems{
+				{
+					Key:   "startup-script",
+					Value: &host.UserData,
+				},
+			},
+		},
+		Scheduling: &compute.Scheduling{
+			AutomaticRestart:  &autoRestart,
+			OnHostMaintenance: "MIGRATE",
+			Preemptible:       false,
+		},
+		Disks: []*compute.AttachedDisk{
+			{
+				AutoDelete: true,
+				Boot:       true,
+				Type:       "PERSISTENT",
+				InitializeParams: &compute.AttachedDiskInitializeParams{
+					DiskName:    host.Name,
+					SourceImage: host.OS,
+					DiskSizeGb:  10,
+					Description: "inlets exit node disk",
+				},
+			},
+		},
+		NetworkInterfaces: []*compute.NetworkInterface{
+			{
+				AccessConfigs: []*compute.AccessConfig{
+					{
+						Type: "ONE_TO_ONE_NAT",
+						Name: "External NAT",
+					},
+				},
+				Network: network,
+			},
+		},
+		ServiceAccounts: []*compute.ServiceAccount{
+			{
+				Email: "default",
+				Scopes: []string{
+					compute.ComputeScope,
+				},
+			},
+		},
+	}
+
+	_, err := p.computeService.Instances.Insert(host.Additional["project_id"], host.Additional["zone"], instance).Do()
+	if err != nil {
+		return nil, err
+	}
+
+	return &ProvisionedHost{
+		ID: constructCustomID(host),
+	}, nil
+}
+
+// Delete removes the GCE instance when the LoadBalancer is deleted
+func (p *GCEProvisioner) Delete(id string) error {
```

Comment:
Do you not also need to delete the firewall rule you created?

Reply:
Sure we can! But I left it out because the firewall rule is only targeting instances with the inlets tag. For the reset of the instances in the project the inlets port remains blocked. Also I was thinking if the user decides to use a tool like inlets-pro he can open up the port just by adding a tag inlets to the instance. Let me know what you think :)